### DISCUSSION TITLE
[Static export] Generated inline scripts, `self.__next_f.push` raises CSP errors

### DISCUSSION THREAD

**Discussion by thisisgit** [vercel/next.js#54152] [App Router] [answered]
Hi, there're some related posts on this issue but opened a new discussion since some are already closed:
* https://github.com/vercel/next.js/discussions/43710
* https://github.com/vercel/next.js/issues/43743
* https://github.com/vercel/next.js/discussions/42170

The error looks like this:
<img width="1081" alt="Screenshot 2023-08-17 at 1 26 28 PM" src="https://github.com/vercel/next.js/assets/13231564/0728b93a-254e-4c20-bfa0-5aa3ac8b1c4d">

And by using CSP generator extension, I can confirm that the error is coming for those `self.__next_f.push` scripts:
<img width="758" alt="Screenshot 2023-08-17 at 1 25 15 PM" src="https://github.com/vercel/next.js/assets/13231564/6f10bb1e-8863-436a-9a61-6d50c5ee0513">

There're solutions for this "only" when using SSR such as using middleware or headers, but it looks like there's none for static export. [This post](https://github.com/vercel/next.js/issues/43743#issuecomment-1557252810) says the only solution for this is using CSP hashes but I can't find a way to add hash to those generated scripts. [This post](https://reesmorris.co.uk/blog/implementing-proper-csp-nextjs-styled-components) shows how to add hash to inline scripts but can't find a way to fetch all those `next_f.push` scripts and apply the hashes.

I even tried to just add `'unsafe-inline'` into `script-src` but as you can see in the first screenshot, hash gets automatically added to CSP (by next.js I guess?) and `unsafe-inline` gets ignored. I added it like this:
<img width="1017" alt="Screenshot 2023-08-17 at 1 37 35 PM" src="https://github.com/vercel/next.js/assets/13231564/98d35a08-ec69-4d31-b18b-aa8b23bd2c0b">

Anyone knows solution to this? 
cc @danieltott 
sorry but I'm tagging you since you looked like an expert to this issue🙏

**Comment by lkleuver** [comment]
I'm developing a static website for a cyber security related organisation. Currently at a crossroads where I need to determine if I need to start rewriting.

I would really prefer to keep using app router since it makes the code a lot cleaner and the site faster.
Is there any input the next team can give regarding static export and strict CSP?
Either a roadmap, an (undocumented) API that can be useful or perhaps a clear and decisive "unsupported"

**Comment by badmintonkid** [comment]
My company also hit the same issue. We are rewriting a portion of our application and generating a static bundle which will be hosted in the larger app with the goal of switching things over in pieces. We aren't using a nonce because we don't want to run next yet and we can't use a hash because the CSP configuration is actually stored in a different repo.

Our solution was to add a postbuild script which extracts the inline scripts from the final HTML files, creates a new asset with the same code, and rewrites the HTML to reference that script. As long as you have `self` allowed for [script-src in your CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/Sources#self), this should work. I'd be open to helping get this into Next.js code if there is interest in going with this approach. If done in Next.js code, it should be much cleaner since we wouldn't need to rewrite the HTML. It's also much less fragile as the regex could break if the inline scripts ever have additional attributes.

I've included the script below. If you aren't storing JS assets in `/assets`, you'll want to change the paths on the rewrite step.

Postbuild script in package.json
```json
"build": "next build",
"postbuild": "node ./scripts/remove-inline-scripts.js",
```

Script
```javascript
const { resolve } = require('path');
const { createHash } = require('crypto');
const { readFileSync, writeFileSync } = require('fs');
const { globSync } = require('glob');

// pretend we are in a production build
process.env.NODE_ENV = 'production';
const { basePath, distDir } = require('../next.config');

const MAGIC_STRING = '__this_is_a_placeholder_for_the_inline_scripts__';

console.log('grab all the html files');
const baseDir = resolve(distDir.replace(/^\//, ''));
const htmlFiles = globSync(`${baseDir}/**/*.html`);
htmlFiles.forEach((file) => {
  // grab inline scripts from each html file
  const contents = readFileSync(file).toString();
  const scripts = [];
  const newFile = contents.replace(/<script>(.+?)<\/script>/g, (_, data) => {
    const addMagicString = scripts.length === 0;
    scripts.push(`${data}${data.endsWith(';') ? '' : ';'}`);
    return addMagicString ? MAGIC_STRING : '';
  });

  // early exit if we have no inline scripts
  if (!scripts.length) {
    return;
  }
  console.log(`processing ${file}`);

  // combine all the inline scripts, add a hash, and reference the new file
  console.log('\trewriting');
  const chunk = scripts.join('');
  const hash = createHash('md5').update(chunk).digest('hex');
  writeFileSync(`${baseDir}/assets/chunk.${hash}.js`, chunk);
  writeFileSync(
    file,
    newFile.replace(
      MAGIC_STRING,
      `<script src="${basePath}${
        basePath.endsWith('/') ? '' : '/'
      }assets/chunk.${hash}.js" crossorigin=""></script>`
    )
  );
  console.log('\tfinished');
});
```

**Reply by badmintonkid** [reply]
I've seen @leerob in a lot of these posts so tagging him for visibility. Static site generation feels incomplete as a feature if we can't also maintain a static CSP. I'm also happy to move this over to https://github.com/vercel/next.js/discussions/54907 if that is a better place for it. It felt like that one was more targeted to using Next as a server.

**Reply by leerob** [reply]
I'd recommend using dynamic rendering for a strict CSP right now. Note, you can have cached data, while still having dynamic rendering. https://demo.vercel.store/ and the associated repo is an example.

**Reply by badmintonkid** [reply]
I can understand using dynamic rendering if you are running Next on a server. But one of the benefits of a static site is that you don't need a server. A strict CSP that only allows scripts from your domain protects against client-side injection attacks making it useful even without a generated nonce.

Possible I'm missing something here when loading a script synchronously instead of running inline scripts. We are still early in our usage of Next.js and there could definitely be use cases I'm not aware of where this approach breaks. A referenced script does result in an additional network call, but that seems like a fair tradeoff to be able to maintain a strict CSP statically. Since we're not using `async` or `defer`, I believe this approach is functionally the same (blocking HTML parsing until the script has been loaded and evaluated).

**Reply by lkleuver** [reply]
no node allowed on our target server, so I've rewritten to Hugo

**Reply by taheramr** [reply]
We are facing a similar issue with this and can't use the headers function as we are using the `export` mode to have a fully static website, will appreciate if there are any clear guidance on how to achieve this somewhere.

**Reply by badmintonkid** [reply]
The script I dropped in my original post runs at build time and moves the inline scripts so that you can remove `unsafe-inline` from the CSP. But you still have to handle the CSP (really all headers) yourself since Next.js is only used to generate the bundle in a static build. For example, this could be done at CDN.
- https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/example-function-add-security-headers.html
- https://cloud.google.com/cdn/docs/web-security-best-practices
- https://learn.microsoft.com/en-us/azure/frontdoor/front-door-security-headers

**Reply by carlosagsmendes** [reply]
> I'd recommend using dynamic rendering for a strict CSP right now. Note, you can have cached data, while still having dynamic rendering. https://demo.vercel.store/ and the associated repo is an example. 

Dynamic implies a performance penalty, and we decided to use Next to improve the responsiveness of our website.

Can Partial Prerendering help? I think of it as a doughnut model were we can have pages pre-rendered and still be able to fill in the holes.

**Comment by leerob** [comment, marked as answer]
- How it works: https://github.com/vercel/next.js/discussions/42170#discussioncomment-8137079
- Docs for `nonce` and CSP: https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy
